Paintroid/src/main/java/org/catrobat/paintroid/iotasks/SaveImageAsync.kt

```diff
@@ -28,102 +29,162 @@ import org.catrobat.paintroid.tools.Workspace
-		fun onSaveImagePostExecute(requestCode: Int, uri: Uri?, saveAsCopy: Boolean)
-		val contentResolver: ContentResolver?
-		val isFinishing: Boolean
-	}
+    @Suppress("Detekt.TooGenericExceptionCaught")
+    override fun doInBackground(vararg params: Void?): Uri? {
+        val callback = callbackRef.get()
+        if (callback == null || callback.isFinishing) {
+            return null
+        }
+        try {
+            val bitmap = workspace.bitmapOfAllLayers
+            val fileName = FileIO.getDefaultFileName()
+            val fileExistsValue = FileIO.checkIfDifferentFile(fileName)
+            return if (FileIO.isCatrobatImage) {
+                val bitmapList = workspace.bitmapLisOfAllLayers
+                return if (uri != null && fileExistsValue == Constants.IS_ORA) {
+                    setUriToFormatUri(fileExistsValue)
+                    uri?.let {
+                        saveOraFile(bitmapList, it, fileName, bitmap, callback.contentResolver)
+                    }
+                } else {
+                    val imageUri =
+                        exportOraFile(bitmapList, fileName, bitmap, callback.contentResolver)
+                    FileIO.currentFileNameOra = fileName
+                    FileIO.uriFileOra = imageUri
+                    imageUri
+                }
+            } else {
+                getImageUri(callback, bitmap)
+            }
+        } catch (e: IOException) {
+            Log.d(TAG, "Can't save image file", e)
+        } catch (e: NullPointerException) {
+            Log.e(TAG, "Can't load image file", e)
+        }
+        return null
+    }
 
-	companion object {
-		private val TAG = SaveImageAsync::class.java.simpleName
-	}
+    private fun setUriToFormatUri(formatcode: Int) {
+        if (formatcode == Constants.IS_JPG) {
+            if (FileIO.uriFileJpg != null) {
+                uri = FileIO.uriFileJpg
+            }
+        } else if (formatcode == Constants.IS_PNG) {
+            if (FileIO.uriFilePng != null) {
+                uri = FileIO.uriFilePng
+            }
+        } else {
+            if (FileIO.uriFileOra != null) {
+                uri = FileIO.uriFileOra
+            }
+        }
+    }
 
-	init {
-		this.uri = uri
-		this.saveAsCopy = saveAsCopy
-		this.workspace = workspace
-	}
-}
\ No newline at end of file
+    override fun onPostExecute(uri: Uri?) {
+        val callback = callbackRef.get()
+        if (callback != null && !callback.isFinishing) {
+            callback.onSaveImagePostExecute(requestCode, uri, saveAsCopy)
+        }
+    }
+
+    interface SaveImageCallback {
+        fun onSaveImagePreExecute(requestCode: Int)
+        fun onSaveImagePostExecute(requestCode: Int, uri: Uri?, saveAsCopy: Boolean)
+        val contentResolver: ContentResolver?
+        val isFinishing: Boolean
+    }
+
+    companion object {
+        private val TAG = SaveImageAsync::class.java.simpleName
+    }
+
+    init {
+        this.uri = uri
```

Comment:
we probably could also make these values, private var/val in the primary constructor, right? so we don't need this init